Client/lib/gMap.js

```diff
@@ -0,0 +1,282 @@
+          icon: {
+            url: "https://s3.amazonaws.com/old.cdn.content/pb/f85b4a1ca1094d22e6c6839e934f048b.png",
+            scaledSize: new google.maps.Size(20,20),
+            origin: new google.maps.Point(0,0),
+            anchor: new google.maps.Point(10,10)
+          }
+      }
+    });
+    //creating waypoints to place on the map
+    var waypoints = [];
+    for (var i = 1; i < gMap.pathLatLng.length-1; i++) {
+      waypoints.push({ location: gMap.pathLatLng[i] });
+    };
+    gMap.directionsService = new google.maps.DirectionsService();
+
+    var request = {
+        origin : gMap.pathLatLng[0],
+        destination : gMap.pathLatLng[gMap.pathLatLng.length-1], // can be latLng or string (this is required)
+        travelMode : google.maps.TravelMode[gMap.travelMode],
+        waypoints : waypoints
+    };
+
+    gMap.directionsService.route(request, function(response, status) {
+      if (status === google.maps.DirectionsStatus.OK) {
+        gMap.directionsDisplay.setDirections(response);
+        gMap.emptyMarkers();
+        gMap.makeMarker(response.routes[0].legs[0].start_location);
+        for (var i = 0; i < response.routes[0].legs.length; i++) {
+          gMap.makeMarker(response.routes[0].legs[i].end_location);
+          gMap.distance[i] = response.routes[0].legs[i].distance.value;
+          gMap.duration[i] = response.routes[0].legs[i].duration.value;
+        };
+        oldPath.setMap(null);
+      }else{
+        gMap.pathLatLng.pop(); //remove the last waypoint added
+      }
+      google.maps.event.addListener(gMap.directionsDisplay, 'directions_changed', function() {
+        if(gMap.directionsDisplay.directions.routes){
+          gMap.pathLatLng[0] = gMap.directionsDisplay.directions.routes[0].legs[0].start_location
+        }
+        for (var i = 1; i <= gMap.directionsDisplay.directions.routes[0].legs.length; i++) {
+          gMap.pathLatLng[i] = gMap.directionsDisplay.directions.routes[0].legs[i-1].end_location
+        };
+        gMap.emptyMarkers();
+        gMap.createPath();
+      });
+    });
+  }
+
+  gMap.makeMarker = function (latLng){
+    var markerImage = {
+      url: 'https://cdn2.iconfinder.com/data/icons/snipicons/500/map-marker-512.png',
+      scaledSize: new google.maps.Size(40,40),
+      origin: new google.maps.Point(0,0),
+      anchor: new google.maps.Point(20,50)
+    };
+
+    var index = gMap.markers.length;
+    var marker = new google.maps.Marker({
+      map: gMap.map,
+      icon: markerImage,
+      position: latLng,
+      title: 'marker ' + index
+    });
+    gMap.markers.push(marker);
+    google.maps.event.addListener(marker, 'click', function() {
+      console.log(index);
+    });
+  }
+  gMap.emptyMarkers = function (){
+    for (var i = 0; i < gMap.markers.length; i++) {
+      gMap.markers[i].setMap(null);
+    };
+    gMap.markers.length = 0;
+  }
+
+  gMap.getGeolocation = function (callback){
+    var pos = new google.maps.LatLng(-33.73, 149.02);
```

Comment:
I still think that passing back an error rather than a random POS is a better option. That way the caller can know that GPS is not available and adjust user expectations.